plugins/image-prioritizer/helper.php

```diff
@@ -137,14 +142,193 @@ function image_prioritizer_add_element_item_schema_properties( array $additional
+				__( 'HEAD request for background image URL did not return an image Content-Type: %s.', 'image-prioritizer' ),
+				$content_type[0]
+			)
+		);
+	}
+
+	/*
+	 * Validate that the Content-Length is not too massive, as it would be better to err on the side of
+	 * not preloading something so weighty in case the image won't actually end up as LCP.
+	 * The value of 2MB is chosen because according to Web Almanac 2022, the largest image by byte size
+	 * on a page is 1MB at the 90th percentile: <https://almanac.httparchive.org/en/2022/media#fig-12>.
+	 * The 2MB value is double this 1MB size.
+	 */
+	$content_length = (array) wp_remote_retrieve_header( $r, 'content-length' );
+	if ( ! is_numeric( $content_length[0] ) ) {
+		return new WP_Error(
+			'background_image_content_length_unknown',
+			__( 'HEAD request for background image URL did not include a Content-Length response header.', 'image-prioritizer' )
+		);
+	} elseif ( (int) $content_length[0] > 2 * MB_IN_BYTES ) {
+		return new WP_Error(
+			'background_image_content_length_too_large',
+			sprintf(
+				/* translators: %s is the content length of the response  */
+				__( 'HEAD request for background image URL returned Content-Length greater than 2MB: %s.', 'image-prioritizer' ),
+				$content_length[0]
+			)
+		);
+	}
+
+	return true;
+}
+
+/**
+ * Sanitizes the lcpElementExternalBackgroundImage property from the request URL Metric storage request.
+ *
+ * This removes the lcpElementExternalBackgroundImage from the URL Metric prior to it being stored if the background
+ * image URL is not valid. Removal of the property is preferable to invalidating the entire URL Metric because then
+ * potentially no URL Metrics would ever be collected if, for example, the background image URL is pointing to a
+ * disallowed origin. Then none of the other optimizations would be able to be applied.
+ *
+ * @since n.e.x.t
+ * @access private
+ *
+ * @phpstan-param WP_REST_Request<array<string, mixed>> $request
+ *
+ * @param WP_REST_Response|WP_HTTP_Response|WP_Error|mixed $response Result to send to the client.
+ *                                                                   Usually a WP_REST_Response or WP_Error.
+ * @param array<string, mixed>                             $handler  Route handler used for the request.
+ * @param WP_REST_Request                                  $request  Request used to generate the response.
+ *
+ * @return WP_REST_Response|WP_HTTP_Response|WP_Error|mixed Result to send to the client.
+ * @noinspection PhpDocMissingThrowsInspection
+ */
+function image_prioritizer_filter_rest_request_before_callbacks( $response, array $handler, WP_REST_Request $request ) {
+	if (
+		$request->get_method() !== 'POST'
+		||
+		// The strtolower() is due to \WP_REST_Server::match_request_to_handler() using case-insensitive pattern match.
+		OD_REST_API_NAMESPACE . OD_URL_METRICS_ROUTE !== trim( strtolower( trim( $request->get_route(), '/' ) ) )
```

Comment:
Indeed, I found that if you tried making a request without pretty permalinks and added `%0A` to the end of the `rest_route`:

```
http://localhost:8888/index.php?rest_route=/optimization-detective/v1/url-metrics:store%0A
```

Then the route is parsed as `"/optimization-detective/v1/url-metrics:store\n"` and this matches here in `\WP_REST_Server::match_request_to_handler()` on [this line](https://github.com/WordPress/wordpress-develop/blob/c697356c29effcfa8115d852bc89385f7112c202/src/wp-includes/rest-api/class-wp-rest-server.php#L1171):

```php
$match = preg_match( '@^' . $route . '$@i', $path, $matches );
```

This is because `$` matches a newline. In reality this should be using `\z` instead.

Reply:
I added a test case to explicitly check for this: https://github.com/WordPress/performance/pull/1713/commits/5ab7fd1a1ca07413795332a89fcf6bb3995360bb